### ISSUE TITLE
almost weekly useful materials - 12/21 -

### ISSUE THREAD

**Issue by GENZITSU** [GENZITSU/UsefulMaterials#121]

**Comment by GENZITSU** [comment]
##  DoWhyとEconMLによる因果推論の実装

Microsoftが開発している因果推論用PythonライブラリDoWhyとEconMLを用いて、どのように因果推論を実装するかを解説したスライド

因果推論そのものについての解説もありとてもわかりやすい。

スライドはこんな感じ

![スクリーンショット 2022-12-07 22 58 19](https://user-images.githubusercontent.com/19774756/206198488-ec2fe04b-a316-4e97-8e17-58038afa0e4f.png)

![スクリーンショット 2022-12-07 22 58 35](https://user-images.githubusercontent.com/19774756/206198500-740147fe-110a-4b73-86e4-88df3c7e8184.png)

![スクリーンショット 2022-12-07 22 58 42](https://user-images.githubusercontent.com/19774756/206198509-2c6164de-d756-45c9-a04d-af1b8ba59608.png)

![スクリーンショット 2022-12-07 22 58 49](https://user-images.githubusercontent.com/19774756/206198527-774881c3-b196-495c-9e21-f2f9118e5eb8.png)

![スクリーンショット 2022-12-07 22 59 01](https://user-images.githubusercontent.com/19774756/206198537-ac2e3946-c8b4-43cf-bc59-9c62cd287ba4.png)

### コメント

著者はzennにて[因果推論に関する解説記事のシリーズ](https://zenn.dev/s1ok69oo)を書かれているのでそちらも参照したい。

### 出典
[DoWhyとEconMLによる因果推論の実装](https://speakerdeck.com/s1ok69oo/dowhytoeconmlniyoruyin-guo-tui-lun-noshi-zhuang)

**Comment by GENZITSU** [comment]
## ChatGPTはどのように学習を行なっているのか

連日界隈を賑わせているChatGPTの学習方法方をその背景から丁寧に説明している記事。

コアとなるのはReinforcement Learning from Human Feedbackというテクニックで、[Deep reinforcement learning from human preferences](https://arxiv.org/abs/1706.03741)というopenAIとgoogleの論文で提案されたものである。

その心は現在の状態価値を判断するpredictorを人間のFBによって学習させるというもので、具体的には2つの行動結果パターンを提示しより価値の高い方を人間に選んでもらうというもの。

![スクリーンショット 2022-12-07 23 23 53](https://user-images.githubusercontent.com/19774756/206204180-341147dd-dbd1-4f44-8fe2-07906feefa2e.png)

![スクリーンショット 2022-12-07 23 23 19](https://user-images.githubusercontent.com/19774756/206203995-d0adbf61-bd13-495c-8b8c-5bbb82b3eedc.png)

この方法は評価者のバイアスや評価者をエージェントが騙してしまうリスクもあるが人間が設定した目的関数よりも良い性能を、少ないステップで到達することが可能。


これを大規模言語モデルに応用したのがInstructGPTで[Training language models to follow instructions with human feedback](https://arxiv.org/abs/2203.02155)というopenAPIの論文で提案されたもの。

ここでは以下のように事前に用意した対話データで学習したモデルにいくつかの回答候補生成してもらい、その回答候補を人間が順位付けしたものを元に報酬予測モデルを学習させるというアプローチをとっている。

![スクリーンショット 2022-12-07 23 30 38](https://user-images.githubusercontent.com/19774756/206205984-eb4af0e7-50ad-4bc3-8356-0743f556b7ac.png)

こうすることで、通常のGPT3よりも性能も高く安全性の高いモデルを作成できるとのこと

![スクリーンショット 2022-12-07 23 33 04](https://user-images.githubusercontent.com/19774756/206206574-9ccfd2e8-c9d4-4394-b5f9-fc869cf1723a.png)

これを対話に特化させたのがchatGPTということになる。

chatGPTにはいくつかの改善点が指摘されているが、これも今後改善さえれていくはずである。


### コメント

人間のFeedBackを強化学習に取り入れる発想が天才的だし、そっちの方が学習が早く進むというのも驚きである。

### 出典
- [ChatGPTはどのように学習を行なっているのか](https://zenn.dev/ttya16/articles/chatgpt20221205)
- [chatGPT](https://openai.com/blog/chatgpt/)

**Comment by GENZITSU** [comment]
## Getting started with Pytorch 2.0 and Hugging Face Transformers

Pytorch 2.0に追加された機能 `torch.compile`を用いてbertの学習が1.2倍ほど高速化するよという紹介記事

huggingfaceでは`TrainingArguments`に`torch_compile=True`を渡すことで実現できる

```
Our absoulte training time went down from 696s to 596s. The train_samples_per_second value increased from 43 to 50. The f1 score is the same/slighty better than the training without torch_compile.
```

### コメント
現在はベータ版だが2023年の前半ごろに正式版がでるとのこと。

Pytorchの公式では51%ほどの向上とされていたが、何か学習の方法ちがうのだろうか

![スクリーンショット 2022-12-09 22 44 43](https://user-images.githubusercontent.com/19774756/206715865-ab90774e-7f08-47d6-b6db-d80b577d7c35.png)


### 出典
- [Getting started with Pytorch 2.0 and Hugging Face Transformers](https://www.philschmid.de/getting-started-pytorch-2-0-transformers)
- [GET STARTED](https://pytorch.org/get-started/pytorch-2.0/)

**Comment by GENZITSU** [comment]
## [Minqi824 / ADBench](https://github.com/Minqi824/ADBench)

NeurIPS2022にて発表されたテーブル・画像・自然言語を含む57種類の異常検知データセットに対して30種類のアルゴリズムの性能を評価した論文の公式レポジトリ

彼らの研究によれば、教師なし学習で常に性能の高い手法はなく、たとえ1%だとしても教師の情報は重要であるというらしい

> - surprisingly none of the benchmarked unsupervised algorithms is statistically better than others, emphasizing the importance of algorithm selection;
> - with merely 1% labeled anomalies, most semi-supervised methods can outperform the best unsupervised method, justifying the importance of supervision;

![スクリーンショット 2022-12-15 19 35 39](https://user-images.githubusercontent.com/19774756/207837524-ffb67659-7ce4-4eab-acb6-10a0c8fdc345.png)

![スクリーンショット 2022-12-15 19 52 09](https://user-images.githubusercontent.com/19774756/207840970-776dc50f-9c2d-4887-9d34-ec8623e4f7da.png)

### コメント

MITライセンスなのが嬉しい

### 出典
- [Minqi824 / ADBench](https://github.com/Minqi824/ADBench)
- [ADBench: Anomaly Detection Benchmark](https://arxiv.org/abs/2206.09426)

**Comment by GENZITSU** [comment]
## 写真から施設名や観光名所を識別する、画像認識モデルとシステム紹介（Yahoo!ブラウザーアプリ）

Yahoo アプリに搭載されているスポット検索で利用されている画像認識モデルの作成方法とシステム構成を紹介しているブログ

### スポット検索とは

![スクリーンショット 2022-12-15 20 12 05](https://user-images.githubusercontent.com/19774756/207844898-33e0af90-6202-4f09-92ea-0db159b589c4.png)

### 構成

> プレフィルター処理、画像の特徴量抽出処理、認識処理の3つで構成

> プレフィルター処理は小規模な二値分類モデルを用いてクエリ画像がランドマーク認識向けの画像かどうかを判定　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　
> 特徴量抽出は特徴量モデルを用いて画像から512次元の特徴量を抽出
> 認識処理では、抽出した特徴量をクエリとして、予め作成してある特徴量インデックスを検索し、検索結果のTop-Kから最終的な認識結果を決定

### プレフィルター

> プレフィルターモデルとして、入力された画像がランドマークかどうかを2値分類するモデルを開発

> 学習データには、ランドマーク画像として[Google Landmarks Dataset v2](https://github.com/cvdfoundation/google-landmark)を抜粋したものと、ランドマークではない画像として内製したデータセットを組み合わせたものを利用

timmのfinetuningにより実装

![スクリーンショット 2022-12-15 23 45 34](https://user-images.githubusercontent.com/19774756/207890466-17294980-32dd-4285-805f-61c3d3477c83.png)

### 特徴抽出

> モデルの学習には[Google Landmarks Dataset v2](https://github.com/cvdfoundation/google-landmark)をベースにクリーニングを実施したデータセットと[Yahoo!ロコ](https://loco.yahoo.co.jp/)の画像等から構成される内製のランドマークデータセットを利用しています。

> モデルtimmのeca_nfnet_l2のpooling層とクラス分類層を外して利用し、CurricularFaceLossを使って学習 (ArcFaceLossよりもよかったとのこと)

![スクリーンショット 2022-12-15 23 49 44](https://user-images.githubusercontent.com/19774756/207891483-233fd690-87b5-429f-bc1a-f6cfe8360bc3.png)

CurricularFaceとは

学習初期は簡単なサンプルに重み付けを行い、後半になるにつれて難しいサンプルに適応的に重み付けをしてくロス

![スクリーンショット 2022-12-16 0 14 52](https://user-images.githubusercontent.com/19774756/207897480-5da12c93-83a1-4e1c-8361-13df8d05acce.png)


### 検索処理

> 対象ランドマーク以外のランドマークが同時に写り込んでいる画像 (東京タワー、富士山など)への対応

> mage Matchingのモデルを用いて他のランドマークが写り込んでいる可能性が高い画像を除去しています。まず重要度が高いと考えられる国内の有名なランドマークを1,000クラス抽出し、それらに対して人手でインデックスする画像を30枚程度選択

> Image Matchingには[kornia](https://github.com/kornia/kornia)の[LoFTR](https://arxiv.org/abs/2104.00680)を利用しています。比較したいくつかの手法の中で、この用途に関しては、LoFTRが精度、速度の両面で優れていました。

### システム構成

NVIDIAのtritron inference serverを利用

ちなみにtritron inference sercerは1つのGPUで複数のモデルを同時に実行できる機能を備えている。

![スクリーンショット 2022-12-15 23 52 16](https://user-images.githubusercontent.com/19774756/207892117-6f5b701a-1403-455f-8c84-93a0d986fe27.png)

### コメント

画像検索システムの作成方法として参考になる。

### 出典
- [写真から施設名や観光名所を識別する、画像認識モデルとシステム紹介（Yahoo!ブラウザーアプリ）](https://techblog.yahoo.co.jp/entry/2022120130379559/)
- [CurricularFace: Adaptive Curriculum Learning Loss for Deep Face Recognition](https://arxiv.org/abs/2004.00288)

**Comment by GENZITSU** [comment]
## [セッショナイズで機械学習モデルの精度を改善した話](https://tech.enigmo.co.jp/entry/2022/12/16/070000)

ユーザーのサイト回遊行動をもとにアイテムのベクトル埋め込みを行うitem2vecの精度を、botのアクセス履歴を除外したことでで向上させた事例の紹介

サイト内の人気ブランドとマイナーブランドの類似度が不可解に高かったことから発覚し、マイナーブランドにアクセスしているユーザーの履歴をみると一定間隔で商品を遷移するbotの存在が判明したとのこと

### コメント

前処理は大事という良い例だが、気づいたのもすごい

### 出典
[セッショナイズで機械学習モデルの精度を改善した話](https://tech.enigmo.co.jp/entry/2022/12/16/070000)

**Comment by GENZITSU** [comment]
## Pythonのウェブフレームワークで実行時間のプロファイリングを取得し可視化する

Flaskに標準搭載されているプロファイラーを用いて、各エンドポイントの処理時間のボトルネックを把握する方法を紹介

![スクリーンショット 2022-12-19 15 46 32](https://user-images.githubusercontent.com/19774756/208364527-9e42a983-945a-4080-9cd5-0725c3527452.png)

[SNAKEVIZ](https://jiffyclub.github.io/snakeviz/)を用いることで、より直感的に可視化することが可能

![スクリーンショット 2022-12-19 15 47 59](https://user-images.githubusercontent.com/19774756/208364673-a268a441-7350-4a2b-ba39-ae1c34441082.png)

### コメント

flask標準にそんな便利なプロファイラーがあるなんて知らなかった。

### 出典
[Pythonのウェブフレームワークで実行時間のプロファイリングを取得し可視化する](https://zenn.dev/yag_ays/articles/e6ad395a5ae3a4)

**Comment by GENZITSU** [comment]
## MoTインターン取り組み報告 ~ 標識検出の高速化 ~

道路標識検出パイプラインの高速化のために入力画像を効率的に縮小する方法を模索している記事

### 入力画像サイズと推論時間の関係

![スクリーンショット 2022-12-20 15 08 33](https://user-images.githubusercontent.com/19774756/208595918-c14f5b0c-e9a8-4c7e-bf60-ff778d19c6b7.png)

### 標識の出現確率に沿って切り取り場所を変えてみる

![スクリーンショット 2022-12-20 15 09 42](https://user-images.githubusercontent.com/19774756/208596083-ba34de70-1828-4dd2-bbce-4ae3dedac523.png)

画像を小さくした時の精度劣化が抑えられることを発見

![スクリーンショット 2022-12-20 15 10 39](https://user-images.githubusercontent.com/19774756/208596193-70cc73dc-4705-4218-9664-8dbfa83afe9e.png)

ただし、他の箇所を切り取ってみると精度が逆に下がるという事象に遭遇

![スクリーンショット 2022-12-20 15 11 30](https://user-images.githubusercontent.com/19774756/208596396-f439b962-c0b0-45d1-8bbe-4d9cdca4cfb3.png)

![スクリーンショット 2022-12-20 15 12 06](https://user-images.githubusercontent.com/19774756/208596411-71057633-9111-42ed-afac-45537ae20af3.png)

### 標識が持つ色相パターンの領域を抽出して画像を縮小

HSV領域で以下のように抽出し、標識がなさそうな部分をカットする
> 青：100 ≦ H ≦ 195、90 ≦ S ≦ 255、100 ≦ V ≦ 255
> 赤：0 ≦ H ≦ 10、64 ≦ S ≦ 255、50 ≦ V ≦ 255

![スクリーンショット 2022-12-20 15 13 58](https://user-images.githubusercontent.com/19774756/208596653-2593d407-5bd3-43bb-ba4d-ac17252bfa5d.png)

色抽出を行う処理に逆に時間がかかる & 精度自体は低下

![スクリーンショット 2022-12-20 15 14 55](https://user-images.githubusercontent.com/19774756/208596764-a9e6bc1b-edf9-45de-a8fa-c30c4e85067d.png)

### アスペクト比を意識した切り取り

Cutmixの要領で一枚の画像から不要な領域を削除し、元のアスペクト比を保つように貼り戻す

![スクリーンショット 2022-12-20 15 16 50](https://user-images.githubusercontent.com/19774756/208596949-933a764b-ad3a-4bbb-ae2f-c66c1a697d4d.png)

精度自体は低下してしまった

![スクリーンショット 2022-12-20 15 18 50](https://user-images.githubusercontent.com/19774756/208597230-2ab58a67-ff16-4f22-a465-5d745fc0b2e9.png)

### コメント

画像サイズの縮小に関してここまで色々アイデアが出せるのがすごい

### 出典
[MoTインターン取り組み報告 ~ 標識検出の高速化 ~](https://lab.mo-t.com/blog/crop_input_images)

**Comment by GENZITSU** [comment]
## 教師なしクラスタリングのその先へ~ 次元削減と定量解析 ~

教師なしクラスタリングによる投手の特徴分析をこれでもかというレベルまで掘り下げている

中でも目を引いたのはあるクラスターのある特徴量の平均値が他のクラスターの平均値とどれくらい異なるかを可視化したvolcano plotという分析手法

![スクリーンショット 2022-12-20 16 09 32](https://user-images.githubusercontent.com/19774756/208604947-5f9e7dc2-fd5b-4911-9bd1-7697e1ebd863.png)

そして特徴量の大きさを比較できるように並べた以下のプロット

![スクリーンショット 2022-12-20 16 13 49](https://user-images.githubusercontent.com/19774756/208605687-10b77e3f-aa0f-4f99-aa16-227ad4f0a97b.png)

### コメント

行われている分析はそこまで高度なものではないが、ここまでクラスタリングの分析を掘り下げている記事も珍しい

### 出典
[教師なしクラスタリングのその先へ~ 次元削減と定量解析 ~](https://zenn.dev/aidemy/articles/f1d3ba3c98b708)

**Comment by GENZITSU** [comment]
## python-vibrato

Viterbiアルゴリズムに基づいた高速な形態素解析器vibratoのpythonバインディングが公開

作者によるとmecab-python3の1.9倍ほど高速に動作するとのこと

```python
# from https://github.com/daac-tools/python-vibrato

import vibrato

with open('path/to/system.dic', 'rb') as fp:
    dict_data = fp.read()
tokenizer = vibrato.Vibrato(dict_data)

tokens = tokenizer.tokenize('社長は火星猫だ')

len(tokens)
#=> 5

list(tokens)
#=> [Token { surface: "社長", feature: "名詞,一般,*,*,*,*,社長,シャチョウ,シャチョー,," },
#    Token { surface: "は", feature: "助詞,係助詞,*,*,*,*,は,ハ,ワ,," },
#    Token { surface: "火星", feature: "名詞,一般,*,*,*,*,火星,カセイ,カセイ,," },
#    Token { surface: "猫", feature: "名詞,一般,*,*,*,*,猫,ネコ,ネコ,," },
#    Token { surface: "だ", feature: "助動詞,*,*,*,特殊・ダ,基本形,だ,ダ,ダ,," }]
```

### コメント

期待

### 出典
- [python-vibrato](https://github.com/daac-tools/python-vibrato)
- [作者紹介ツイート](https://twitter.com/vbkaisetsu/status/1602939326513700865?s=20&t=EZHjaHER3mfKIh0IFTrF2Q)

**Comment by GENZITSU** [comment]
## NLP Tricks Map 2022

クニえもんさんがNLPコンペに頻出するテクニック達を一つに図にまとめたもの

![Fi4d0qjVEAAhnhv](https://user-images.githubusercontent.com/19774756/208681291-8e0ef9c2-5f2f-4861-8c20-edfd6a63dd13.jpeg)


初耳だったのは以下
- Meta Pseudo Labels ([参考](https://qiita.com/takoroy/items/08ce473d304b8f987c32))
    - validation損失を改善する方向にpseudo labelの出し方を学習させる手法
- Stochastic Weight Averaging ([参考](https://pytorch.org/blog/stochastic-weight-averaging-in-pytorch/))
    - 学習後期に得られた重みの平均を用いて、最終的な重みとする方法
    - pytorchによる実装ではcyclicにlearning rateを上下させて平均を得ることもできる
- convert checkpoint to lsg ([参考](https://github.com/ccdv-ai/convert_checkpoint_to_lsg))
    - 長い系列を扱うためにattentionをいじるテクニックのこと

### コメント

めちゃ網羅的に整理されていて良い

### 出典
[元ツイート](https://twitter.com/kun1em0n/status/1598247713967333382?s=20&t=EZHjaHER3mfKIh0IFTrF2Q)
